# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [8]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [10]:
step3_output = pd.read_csv(r"C:/Users/Curtis/Documents/SpringboardGuidedCapstone/data/step2_output.csv")
#dfno['cluster'] = y_kmeans3
step3_outputdf = step3_output.drop(step3_output.columns[1], axis=1)
step3_outputdf.head()

Name    state  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort   Alaska         3939           2500        250      1   
1  Eaglecrest Ski Area   Alaska         2600           1540       1200      0   
2     Hilltop Ski Area   Alaska         2090            294       1796      0   
3     Arizona Snowbowl  Arizona        11500           2300       9200      0   
4  Sunrise Park Resort  Arizona        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  ...  LongestRun_mi  \
0          0          0          2     2  ...              1   
1          0          0          0     0  ...              2   
2          0          0          0     0  ...              1   
3          0          1          0     2  ...              2   
4          0          0          1     2  ...              1   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               1610             113               150         60   
1                640              60                45         44   
2                 30              30               150         36   
3                777             104               122         81   
4                800              80               115         49   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0              669            65            85                150   
1              350            47            53                 90   
2               69            30            34                152   
3              260            89            89                122   
4              250            74            78                104   

   NightSkiing_ac  
0             550  
1               0  
2              30  
3               0  
4              80  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [11]:
print(step3_outputdf.head())

                  Name    state  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort   Alaska         3939           2500        250      1   
1  Eaglecrest Ski Area   Alaska         2600           1540       1200      0   
2     Hilltop Ski Area   Alaska         2090            294       1796      0   
3     Arizona Snowbowl  Arizona        11500           2300       9200      0   
4  Sunrise Park Resort  Arizona        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  ...  LongestRun_mi  \
0          0          0          2     2  ...              1   
1          0          0          0     0  ...              2   
2          0          0          0     0  ...              1   
3          0          1          0     2  ...              2   
4          0          0          1     2  ...              1   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               1610             113               150     

In [12]:
#dfo = step3_output.select_dtypes(include=['object']) # select object type columns
#step3_output = pd.concat([step3_output.drop(dfo, axis=1), pd.get_dummies(dfo)], axis=1)

dfo = pd.concat([step3_outputdf, pd.get_dummies(step3_outputdf['state'])], axis=1).drop(['state'], axis =1)
print(dfo.shape)
dfo.head()

#dfo = pd.get_dummies(step3_outputdfno, prefix=['state'], columns=['state'])

(330, 60)


Name  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort         3939           2500        250      1   
1  Eaglecrest Ski Area         2600           1540       1200      0   
2     Hilltop Ski Area         2090            294       1796      0   
3     Arizona Snowbowl        11500           2300       9200      0   
4  Sunrise Park Resort        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  triple  ...  Rhode Island  \
0          0          0          2     2       0  ...             0   
1          0          0          0     0       0  ...             0   
2          0          0          0     0       1  ...             0   
3          0          1          0     2       2  ...             0   
4          0          0          1     2       3  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 60 columns]

In [13]:
print(dfo.head())

                  Name  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort         3939           2500        250      1   
1  Eaglecrest Ski Area         2600           1540       1200      0   
2     Hilltop Ski Area         2090            294       1796      0   
3     Arizona Snowbowl        11500           2300       9200      0   
4  Sunrise Park Resort        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  triple  ...  Rhode Island  \
0          0          0          2     2       0  ...             0   
1          0          0          0     0       0  ...             0   
2          0          0          0     0       1  ...             0   
3          0          1          0     2       2  ...             0   
4          0          0          1     2       3  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1     

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [14]:
# first we import the preprocessing package from the sklearn library
from sklearn.preprocessing import StandardScaler

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = dfo.drop(columns=['Name','AdultWeekend'], axis=0)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y= dfo['AdultWeekend']
# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [15]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [16]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [17]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [18]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [19]:
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error

In [20]:
explained_variance_score(y_test, y_pred)

0.934952599733997

In [21]:
mean_absolute_error(y_test, y_pred)

5.097342511478456

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [22]:
lm.intercept_

64.08828441913312

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [23]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)

Coefficient
total_chairs       1.207998e+13
fastQuads          4.579548e+12
surface            4.290693e+12
double             3.781118e+12
triple             3.373017e+12
quad               2.733705e+12
fastSixes          1.357609e+12
trams              1.166496e+12
New York           3.620200e+11
Michigan           3.416486e+11
Colorado           3.010122e+11
California         2.945685e+11
Pennsylvania       2.810958e+11
New Hampshire      2.591924e+11
Wisconsin          2.591924e+11
Vermont            2.513612e+11
Minnesota          2.432232e+11
Utah               2.347463e+11
Idaho              2.258924e+11
Montana            2.258924e+11
Massachusetts      2.166153e+11
Washington         2.068580e+11
Oregon             2.068580e+11
Maine              1.965492e+11
New Mexico         1.965492e+11
Wyoming            1.855967e+11
North Carolina     1.612299e+11
Connecticut        1.474090e+11
Ohio               1.474090e+11
West Virginia      1.320493e+11
Virginia           1.320493e+11
Illinois           1.320493e+11
Nevada             1.320493e+11
fastEight          1.146780e+11
Alaska             1.145333e+11
Iowa               1.145333e+11
New Jersey         9.365896e+10
Missouri           9.365896e+10
Indiana            9.365896e+10
Arizona            9.365896e+10
South Dakota       9.365896e+10
Tennessee          6.632776e+10
Maryland           6.632776e+10
Rhode Island       6.632776e+10
AdultWeekday       1.928320e+01
summit_elev        1.791593e+01
base_elev          1.723222e+01
vertical_drop      5.777534e+00
Runs               1.135986e+00
NightSkiing_ac     7.806091e-01
LongestRun_mi      5.087891e-01
daysOpenLastYear   4.514160e-01
Snow Making_ac     3.765259e-01
yearsOpen          3.738098e-01
averageSnowfall    2.751465e-01
SkiableTerrain_ac  2.050781e-01
projectedDaysOpen  1.873169e-01
TerrainParks       1.302490e-01

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [41]:
# first we import the preprocessing package from the sklearn library
from sklearn.preprocessing import StandardScaler

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = step3_outputdf.drop(columns=['Name','AdultWeekend', 'state'], axis=0)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y= step3_outputdf['AdultWeekend']
# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [42]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [43]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [44]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

In [45]:
explained_variance_score(y_test, y_pred)

0.9249457317984838

In [46]:
mean_absolute_error(y_test, y_pred)

5.509465771878881

In [47]:
lm.intercept_

64.02212244694796

In [48]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)

Coefficient
AdultWeekday         20.214134
summit_elev          12.826668
base_elev             9.926771
vertical_drop         4.532261
averageSnowfall       1.889312
quad                  1.543795
triple                1.358135
surface               1.263800
daysOpenLastYear      1.193342
Runs                  1.089008
fastQuads             0.775560
total_chairs          0.698888
projectedDaysOpen     0.492843
fastSixes             0.429307
LongestRun_mi         0.411744
NightSkiing_ac        0.374663
double                0.298441
SkiableTerrain_ac     0.292680
yearsOpen             0.287808
trams                 0.167608
TerrainParks          0.162943
Snow Making_ac        0.076362
fastEight             0.031734

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [49]:
# first we import the preprocessing package from the sklearn library
from sklearn.preprocessing import StandardScaler

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = step3_outputdf.drop(columns=['Name','AdultWeekend', 'state', 'summit_elev', 'base_elev'], axis=0)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y= step3_outputdf['AdultWeekend']
# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [50]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [51]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [52]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

In [53]:
explained_variance_score(y_test, y_pred)

0.9264111468885956

In [54]:
mean_absolute_error(y_test, y_pred)

5.463404236478459

In [55]:
lm.intercept_

64.02188424123845

In [56]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)

Coefficient
AdultWeekday         20.134047
averageSnowfall       2.241683
quad                  1.587639
triple                1.325705
Runs                  1.289670
surface               1.249296
vertical_drop         1.217295
daysOpenLastYear      1.063306
fastQuads             0.863139
total_chairs          0.745321
NightSkiing_ac        0.517925
fastSixes             0.433358
SkiableTerrain_ac     0.376808
LongestRun_mi         0.370306
double                0.307680
yearsOpen             0.307456
projectedDaysOpen     0.257841
trams                 0.142568
fastEight             0.120909
TerrainParks          0.078313
Snow Making_ac        0.050752

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.9345 | 5.097 |-|
| Model 2. | 0.9249| 5.509 |'state'|
| Model 3. | 0.9264 | 5.463 |'state','summit_elev','base_elev'|

Model Selection: Model 3